### PR TITLE
[MAINTENANCE] use invoke for common contrib/dev tasks

### PR DESCRIPTION
**Pull request by Kilo59**
## Changes proposed in this pull request:

~Combing a few developer/contributor quality of life improvements in this PR, but we may want to break these up into separate PRs before merging.~

I've moved anything not related to the `invoke` task file to other PR will be opening those shortly.

1. ~Creating a [`pyproject.toml` file](https://peps.python.org/pep-0518/) to define the build system ge uses and start centralizing tool config to it (black, isort, pytest, etc.)]~ moved -> https://github.com/great-expectations/great_expectations/pull/5525

2. Begin to use the [`invoke` python command line task runner](https://www.pyinvoke.org/) for running common developer tasks. Define a set of starter tasks like linting, formatting, type checking, etc. These can serve as a guide for future tasks. By providing easy to use and understand generic command line tasks like `lint`, `fmt` etc. we make it easy for new contributors to get up and running quickly without having to know anything about the details of our particular tooling. This also makes it easier to change the underlying tools and/or configuration without breaking anyones workflow. 
But contributors are still free to use each of these tools directly.
Alternatives to `invoke`: Make, [duty](https://github.com/pawamoy/duty), [Pipfile scripts](https://pipenv-fork.readthedocs.io/en/latest/advanced.html#custom-script-shortcuts).

4. ~Building on the [excellent work](https://github.com/great-expectations/great_expectations/blob/develop/scripts/check_type_hint_coverage.py) of  @cdkini begin to fully type-check select ge modules.~ moved -> #5540 

5. ~[Enforce our pre-commit hooks in the CI](https://pre-commit.com/#usage-in-continuous-integration) (WIP) - [use pre-commit ci instead of azure?
](https://pre-commit.ci/)~ moved -> #5526 

image

### TLDR Summary 

- ~add pyproject.toml file and move tool config to it~
  - ~move isort config~
  - ~move pytest config~
- [x]  add `invoke` task runner and define common developer tasks
  - [x]  invoke fmt
  - [x]  invoke sort
  - [x]  invoke lint
  - [x]  invoke hooks (run pre-commit hooks)
  - [x]  invoke types (type-check?)
  - [x] update CI to use new invoke tasks
- ~add check-json pre-commit hook~
- ~[update CI to enforce pre-commit hooks](https://pre-commit.com/#usage-in-continuous-integration)~
- ~begin "real" mypy static type checking of select modules~
  - ~fix all typing errors on select modules~
- [x] [update contrib docs to reflect changes](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style#linting)

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- ~I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.~
- ~I have run any local integration tests and made sure that nothing is broken.~

### CODE REVIEW

**Comment by Kilo59** [comment]
@cla-bot check

File: pyproject.toml
Comment (anthonyburdi) [inline pyproject.toml]: Love this. FYI I think there is or is soon to be a `unit` marker from Bill's recent PR cc @billdirks

File: pyproject.toml
Comment (Kilo59) [inline pyproject.toml:8]: Need to verify that this still works OUTSIDE of a virtual environment.
I know isort made some improvements to the way it detects these packages recently, but some of this may still be needed.

File: tasks.py
Comment (Kilo59) [inline tasks.py]: Might change this to `type_check`, which would be called like -> `invoke type-check`.
It feels more natural but it's more typing 😆 than `invoke types`

File: pyproject.toml
Comment (billdirks) [reply pyproject.toml]: See https://github.com/great-expectations/great_expectations/pull/5481

File: tasks.py
Comment (billdirks) [inline tasks.py]: I like this incremental ability to add modules. What are the numbers after the #, the number of current errors?

File: tasks.py
Comment (billdirks) [inline tasks.py]: I'm wondering if we should just have 2 separate tasks, one for mypy and one `check_type_hint_coverage`. Some of the arguments like `modules` are ignored by `check_type_hint_coverage`.

File: pyproject.toml
Comment (billdirks) [inline pyproject.toml]: How/where is this file read in?

**Review by billdirks — COMMENTED** [review]
This looks nice! I have some questions inline. Also, since you are a core developer you can clone and branch in the main repo instead of forking it.

File: tasks.py
Comment (Kilo59) [reply tasks.py]: Exactly.
I plan to address all the errors for all the un-commented modules before merging, and then add the this check to the CI to prevent regressions.

File: pyproject.toml
Comment (Kilo59) [reply pyproject.toml]: This gets read by pip, also thanks for the question. It made me realize I need to add `wheel` here.
https://peps.python.org/pep-0518/#build-system-table

File: tasks.py
Comment (Kilo59) [reply tasks.py]: Yeah initially had the same thought. I'm open to this change but let me explain my thinking first.

My idea was that the `coverage` check is like a `type-checking-light`, and eventually (possibly many months from now) we will stop using it and exclusively use standard `mypy` type-checking. My thinking is that having two separate type-checking commands would lead to confusion.

File: tasks.py
Comment (billdirks) [reply tasks.py]: Maybe add a comment stating this. They might fall out of date but that's ok.

File: tasks.py
Comment (billdirks) [reply tasks.py]: I can see it both ways. I think 2 distinct steps is clearer (maybe with names like `type-check`, `type-coverage`). We can then wholesale delete the `type-coverage` task when we are ready and don't have to edit this function. I don't feel strongly about this though, either way is fine with me.

File: pyproject.toml
Comment (Kilo59) [reply pyproject.toml]: Added

File: tasks.py
Comment (Kilo59) [reply tasks.py]: Done. Also I changed it to `packages` instead of `modules`.

File: pyproject.toml
Comment (Kilo59) [reply pyproject.toml]: @billdirks moved this work to https://github.com/great-expectations/great_expectations/pull/5525

Comment (Kilo59) [reply pyproject.toml]: moved to #5525

File: tasks.py
Comment (Kilo59) [reply tasks.py]: @billdirks I went with the approach of having 2 separate tasks.
`invoke type-check` & `invoke type-coverage`.
I'll be actually making the initial typing fixes (for our modules with low error counts) in a different PR.

File: docs/contributing/style_guides/code_style.md
Comment (Kilo59) [inline docs/contributing/style_guides/code_style.md]: These steps are no longer necessary as of `isort` 5 and is the reason tools like  `seed-isort` were deprecated.
https://github.com/asottile-archive/seed-isort-config#deprecated

https://pycqa.github.io/isort/docs/configuration/pre-commit.html#seed-isort-config

File: docs/contributing/style_guides/code_style.md
Comment (cdkini) [inline docs/contributing/style_guides/code_style.md]:
```suggestion
  upgrade         Run code syntax upgrades.
```

File: great_expectations/core/usage_statistics/package_dependencies.py
Comment (cdkini) [inline great_expectations/core/usage_statistics/package_dependencies.py:81]: Let's double check with @anthonyburdi that we want this here - I believe this list is for what we want to track through our usage stats.

File: tasks.py
Comment (cdkini) [inline tasks.py:45]: reminds me of the lovely `go fmt`

File: tasks.py
Comment (cdkini) [inline tasks.py]:
```suggestion
    """Run code syntax upgrades."""
```

File: tasks.py
Comment (cdkini) [inline tasks.py]: This is bare `mypy` right? Do we need to add it to our deps? What about a config file?

File: tasks.py
Comment (Kilo59) [inline tasks.py:45]: Wanted to call this out, as opinions here my vary.
By default I have the `invoke fmt` command also run `isort` unless you explicitly tell it not to.

opt-out vs opt-in.

I'm open to either.

Comment (Kilo59) [reply tasks.py:45]: go fmt is definitely the inspiration for this 😄

File: great_expectations/core/usage_statistics/package_dependencies.py
Comment (anthonyburdi) [reply great_expectations/core/usage_statistics/package_dependencies.py:81]: That is correct - it should be in this list but also in `GE_DEV_DEPENDENCIES_EXCLUDED_FROM_TRACKING` since we won't want to track the version used for invoke and other development related packages (e.g. black, isort).

File: tasks.py
Comment (Kilo59) [reply tasks.py]: Yes, on all counts.
My plan was to include all that in a fast-follow that has that plus typing fixes to all these low error count modules + adding this full type-checking to the CI (I have that step commented out in the azure-pipeline yaml).

I could also just strip this `invoke type-check` task out entirely as part of this PR and add it back for the `type-checking` PR.
☝️ I just went ahead and did this.

I'd also like to put the mypy config in pyproject.toml so in terms of ordering these prs.

invoke stuff -> [pyproject.toml stuff](https://github.com/great-expectations/great_expectations/pull/5525) -> (mypy stuff | pre-commit ci stuff)

File: great_expectations/core/usage_statistics/package_dependencies.py
Comment (Kilo59) [reply great_expectations/core/usage_statistics/package_dependencies.py:81]: Done

File: tasks.py
Comment (cdkini) [reply tasks.py]: works for me!

File: tasks.py
Comment (cdkini) [reply tasks.py:45]: I like it! Easy enough to modify if we want to later on

**Review by cdkini — APPROVED** [review]

**Review by talagluck — APPROVED** [review]
LGTM!

**Comment by AFineDayFor** [comment]
little nit and stretch thing, leveraging click to manage args to invoke?

**Comment by Kilo59** [comment]
@AFineDayFor 
>little nit and stretch thing, leveraging click to manage args to invoke?

As discussed, the intended use-cases for click and `invoke` are a bit different.
[Invoke](https://www.pyinvoke.org/) is meant to be more of a light-weight, local, Make-like task runner, while `click` is more of heavier CLI application framework.

Although [invoke can be used in the same way](https://docs.pyinvoke.org/en/stable/concepts/library.html#reusing-invoke-s-cli-module-as-a-distinct-binary) and click CAN be used to run common dev tasks.

**Review by Shinnnyshinshin — APPROVED** [review]
Love it... thank you very much for introducing these changes @Kilo59

**Pull request merged**
